atorch/examples/ds_3d_parallel/ds_3d_llama2.py

```diff
@@ -0,0 +1,168 @@
+import argparse
+import datetime
+
+import torch
+from example_utils import compute_llama2_training_flops, get_data_iter, print_rank_0, report_memory, sync_and_time
+from transformers import AutoConfig, AutoTokenizer
+from transformers.models.llama.modeling_llama import LlamaModel
+
+import atorch
+from atorch.auto import auto_accelerate
+from atorch.auto.opt_lib.ds_3d_parallel_optimization import DeepSpeed3DParallelConfig
+from atorch.common.util_func import divide
+from atorch.modules.distributed_modules.cross_entropy import vocab_parallel_cross_entropy
+from atorch.utils.manual_tp_utils import TPInfo
+from atorch.utils.meta_model_utils import record_module_init
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Pretrain llama2 with atorch ds 3d.")
+    parser.add_argument(
+        "--ds_config",
+        type=str,
+        default=None,
+        help="Deepspeed config json file path.",
+    )
+    parser.add_argument(
+        "--model_name_or_path",
+        type=str,
+        help="Path to pretrained model or model identifier from huggingface.co/models.",
+        required=False,
+    )
+    parser.add_argument(
+        "--dataset_path",
+        type=str,
+        default=None,
+        help="A dir containing dataset with .arrow format.",
+    )
+    parser.add_argument(
+        "--block_size",
+        type=int,
+        default=None,
+        help=(
+            "Optional input sequence length after tokenization. The training dataset will be truncated in block of"
+            " this size for training. Default to the model max input length for single sentence inputs (take into"
+            " account special tokens)."
+        ),
+    )
+    parser.add_argument(
+        "--model_parallel_size", type=int, default=1, help="size of the model parallel. specific for tensor parallel"
+    )
+    parser.add_argument("--pipeline_parallel_size", type=int, default=1, help="size of the pipeline parallel.")
+    args = parser.parse_args()
+
+    return args
+
+
+def get_llama_tpinfo():
+    llama_tpinfo = TPInfo()
+    llama_tpinfo.shard_col("self_attn.q_proj", "self_attn.k_proj", "self_attn.v_proj", "mlp.gate_proj", "mlp.up_proj")
+    llama_tpinfo.shard_row("self_attn.o_proj", "mlp.down_proj")
+    llama_tpinfo.shard_vocab("embed_tokens")
+    llama_tpinfo.shrink({"self_attn": {"num_heads", "num_key_value_heads", "hidden_size"}})
+    return llama_tpinfo
+
+
+def get_llama_3d_parallel_cfg(args):
+    def batch_fn(data):
+        tokens = data["input_ids"]
+        position_ids = data["position_ids"]
+        attention_mask = data["attention_mask"]
+        labels = data["labels"]
+        loss_mask = data["loss_mask"]
+
+        return (tokens, attention_mask, position_ids), (labels, loss_mask)
+
+    ds_config = args.ds_config
```

Comment:
Is this args.ds_config required? Or just leave it as default None?

Reply:
https://github.com/intelligent-machine-learning/dlrover/blob/ds_3d_llama2/atorch/examples/ds_3d_parallel/llama2_entry.sh#L54